Fix: Update Safe Mode notice to be consistent with new changes from Jetpack

htdat opened the pull request:
Fixes #9614

There were a few consistencies between Jetpack and WooPayments safe-mode notices due to: 

- Bug in Jetpack when using the callback to get custom texts https://github.com/Automattic/woocommerce-payments/pull/9884 - [details](https://github.com/Automattic/woocommerce-payments/pull/10489#discussion_r1984410626). ✅  This was fixed in https://github.com/Automattic/jetpack/pull/42265
- Jetpack introduces another type of Safe Mode with new custom texts, specifically for sites in the development mode. There were bugs in introducing these new texts but they were fixed in https://github.com/Automattic/jetpack/pull/42235 and https://github.com/Automattic/jetpack/pull/42280

> [!CAUTION]
>  TODOs: 
> - [x] Update jeptack-connection to the latest version, and ensure all fixing PRs are in this version.


#### Changes proposed in this Pull Request

- Provide WooPayments-specific wordings for the new custom texts of the Safe Mode notice for the dev env.
- Fix the issue of reversed URLs. 

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- you can test the IDC behavior and the Safe Mode notice by installing this https://jetpack.com/download-jetpack-beta/, then activating `jetpack-debug-helper`, then `Identity Crisis Simulation Utility`. 
- even with the IDC simulator, you still need to test this in your non-local host address by using ngrok or Jurassic Tube. 
- to trigger between two types of notices, use this constant in wp-config.php, and chang value to either `development` or `production`:  `define( 'WP_ENVIRONMENT_TYPE', 'development' );` 
- expect: see two different notices.
- expect: in the two notices, all text contents and links are WooPayments-relevant.

#### Screenshot - site in Production Mode:

![isDev-false](https://github.com/user-attachments/assets/87fa976d-dcfa-4419-ad61-2d14faa1dc8d)

#### Screenshot - site in Dev Mode, this is a new Safe Mode notice with a slightly different descriptions and button labels: 

![isDev-true](https://github.com/user-attachments/assets/eba64a10-0db5-497d-a603-256de638b070)

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

> [!IMPORTANT]
> After this PR is mergerd, ensure that https://woocommerce.com/document/woopayments/testing-and-troubleshooting/safe-mode/ is updated.  

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [x] No need - Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

htdat committed:
Update texts

based on PR 10489

htdat committed:
Merge branch 'develop' into fix/9614-jetpack-safemode-notice

htdat committed:
Changelog

- minor typo

htdat committed:
Add mainBodyTextDev custom text

See https://github.com/Automattic/jetpack/pull/42280

htdat committed:
Merge branch 'develop' into fix/9614-jetpack-safemode-notice

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `10511` or branch name `fix/9614-jetpack-safemode-notice` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/9614-jetpack-safemode-notice) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: e45b617897840344968b3fb52b55bad759d701f1
- Build time: 2025-03-11 12:49:34 UTC

_Note: the build is updated when a new commit is pushed to this PR._

htdat committed:
Merge branch 'develop' into fix/9614-jetpack-safemode-notice

htdat committed:
This actually reverts PR 8033 as we no longer need this anymore.

Looking at [Jetpack code](https://github.com/Automattic/jetpack/blob/06cda771335eca520400af8dd2c13a3769b75e9f/projects/packages/connection/src/identity-crisis/class-identity-crisis.php#L501-L506), we expect that Jetpack should return reversed URLs but it seems Jetpack is no longer do so.

htdat commented on woocommerce-payments.php:
```diff
@@ -355,16 +356,7 @@ function wcpay_get_jetpack_idc_custom_content(): array {
 	$urls = Automattic\Jetpack\Identity_Crisis::get_mismatched_urls();
 	if ( false !== $urls ) {
 		$current_url = untrailingslashit( $urls['current_url'] );
-		/**
-		 * Undo the reverse the Jetpack IDC library is doing since we want to display the URL.
-		 *
-		 * @see https://github.com/Automattic/jetpack-identity-crisis/blob/trunk/src/class-identity-crisis.php#L471
-		 */
-		$idc_sync_error = Automattic\Jetpack\Identity_Crisis::check_identity_crisis();
-		if ( is_array( $idc_sync_error ) && ! empty( $idc_sync_error['reversed_url'] ) ) {
-			$urls['wpcom_url'] = strrev( $urls['wpcom_url'] );
-		}
-		$wpcom_url = untrailingslashit( $urls['wpcom_url'] );
```
This actually reverts PR #8033 as we no longer need this anymore. 

Looking at [Jetpack code](https://github.com/Automattic/jetpack/blob/06cda771335eca520400af8dd2c13a3769b75e9f/projects/packages/connection/src/identity-crisis/class-identity-crisis.php#L501-L506), we expect that Jetpack should return reversed URLs but it seems Jetpack is no longer do so. Would love to have your confirmation if this is the correct way to doing so now @bindlegirl and/or @Automattic/jetpack-vulcan 

Without this revert, the reversed URL is being used for the original site address: 

![Markup on 2025-03-10 at 10:51:26](https://github.com/user-attachments/assets/57e6e5e0-2a21-4002-b467-6eeefe516cc9)

bindlegirl replied:
The reversal code was the same since it was introduced so what users reported for WCPay and reverted URLs shouldn't be possible.
Could it be that there was a conflict of IDC versions? For example, users installed a newer Jetpack that had the reversing logic that wrote the IDC domains in reverse, then they disabled it, and what was left was WCPay with an older version of the IDC class with no reversal, so they experienced the strange reversed URL?

Maybe the fix on WCPay side should include the package version check instead of removing it completely?

htdat replied:
> Could it be that there was a conflict of IDC versions? For example, users installed a newer Jetpack that had the reversing logic that wrote the IDC domains in reverse, then they disabled it, and what was left was WCPay with an older version of the IDC class with no reversal, so they experienced the strange reversed URL?

That could be but I am not entirely sure. 

@oaratovskyi @vladolaru @ismaeldcom - could you have a look at this specific part? For now, removing your fix on #8033 helped to fix the issue; otherwise, now it becomes this issue #8025

ismaeldcom replied:
🤔  That's strange, IIRC the fix should only reverse an already reversed URL.

bindlegirl replied:
I was able to confirm my theory. You can access a test JN site using this data: 3732b-pb/

I loaded an old WCPay version (pre custom WCPay fix) with a new Jetpack. Since Jetpack created the Safe Mode reversed entry, the old WCPay wasn't able to work with that and displayed a reversed URL. Try dis/re-enabling WCPay on the JN test site for results.
So your fix should only run on WCPay when using identity-crisis package pre-v0.10.6 (see here: https://github.com/Automattic/jetpack-identity-crisis/compare/v0.10.5...v0.10.6). If the connection package is used for idc (since we moved the idc package to connection one), you should be fine.

htdat replied:
@bindlegirl and @ismaeldcom - thanks for providing more info and the test! If so, the right thing here is just to delete the patch created by PR #8033, i.e. what I am doing it now.

oaratovskyi replied:
@htdat yes, that was a fix because Jetpack was sending reversed URLs to us in specific cases. If it doesn't anymore, we could delete the patch.

htdat committed:
Merge branch 'develop' into fix/9614-jetpack-safemode-notice

htdat committed:
update jetpack-connection to 6.7.3

htdat committed:
Merge branch 'develop' into fix/9614-jetpack-safemode-notice

dmvrtx reviewed (approved):
Thanks for working on this, @htdat! I tested in both environment modes, and the message itself and proposed actions are top-notch!
